google-cloud-pubsub/lib/google/cloud/pubsub/subscriber.rb

```diff
@@ -0,0 +1,386 @@
+        def initialize callback, subscription_name, deadline, inventory,
+                       threads, service
+          @request_queue = nil
+          @output_enum = nil
+          @callback = callback
+          @subscription_name = subscription_name
+          @deadline = deadline || 60
+          @inventory = Inventory.new self, (inventory || 100)
+          @threads = threads || [2, Concurrent.processor_count * 2].max
+          @service = service
+
+          super() # to init MonitorMixin
+        end
+
+        def start
+          synchronize do
+            return if @request_queue
+
+            @request_queue = EnumeratorQueue.new self
+            @thread_pool = Concurrent::FixedThreadPool.new threads
+            start_streaming!
+          end
+
+          true
+        end
+
+        def stop
+          synchronize do
+            return if @request_queue.nil?
+            @request_queue.push self
+            @inventory.stop
+            @stopped = true
+          end
+
+          true
+        end
+
+        def stopped?
+          synchronize { @stopped }
+        end
+
+        def paused?
+          synchronize { @paused }
+        end
+
+        def wait!
+          synchronize do
+            @background_thread.join if @background_thread
+
+            @thread_pool.shutdown
+            @thread_pool.wait_for_termination
+          end
+
+          true
+        end
+
+        ##
+        # @private
+        def acknowledge *messages
+          ack_ids = coerce_ack_ids messages
+          return true if ack_ids.empty?
+
+          ack_request = Google::Pubsub::V1::StreamingPullRequest.new
+          ack_request.ack_ids += ack_ids
+
+          synchronize do
+            @request_queue.push ack_request
+            @inventory.remove ack_ids
+            unpause_streaming!
+          end
+
+          true
+        end
+
+        ##
+        # @private
+        def delay new_deadline, *messages
+          deadline_ack_ids = coerce_ack_ids messages
+          return true if deadline_ack_ids.empty?
+
+          deadline_seconds = deadline_ack_ids.count.times.map { new_deadline }
+          deadline_ack_request = Google::Pubsub::V1::StreamingPullRequest.new
+          deadline_ack_request.modify_deadline_ack_ids += deadline_ack_ids
+          deadline_ack_request.modify_deadline_seconds += deadline_seconds
+
+          synchronize do
+            @request_queue.push deadline_ack_request
+            @inventory.remove deadline_ack_ids
+            unpause_streaming!
+          end
+
+          true
+        end
+
+        def inventory
+          synchronize { @inventory.count }
+        end
+
+        ##
+        # @private
+        def delay_inventory!
+          synchronize do
+            return true if @inventory.empty?
+
+            ack_ids_deadline = @inventory.ack_ids.map { deadline }
+            inventory_ack_request = Google::Pubsub::V1::StreamingPullRequest.new
```

Comment:
I would make sure that inventory does not exceed 10 MB. This should be unlikely, but if it does you want to split it into multiple StreamingPullRequests.